*: Support reviewing audiences of token

brancz opened the pull request:
This adds the ability to verify the audience of a token, so clients can use audience scoped tokens, as opposed to essentially allowing the kube-rbac-proxy to impersonate the requesting client using its serviceaccount token.

Closes #18 

@s-urbaniak @paulfantom @LiliC @metalmatze

s-urbaniak commented:
do you mind to put the vendoring into a separate commit?

brancz commented:
Done.

s-urbaniak commented on examples/non-resource-url-token-request/non-resource-url/README.md:
```diff
@@ -0,0 +1,170 @@
+# non-resource-url example
```
hmm, how does this differ from `examples/non-resource-url-token-request/README.md`? Or is this duplication just accidentally?

brancz replied:
whops I haven't finished this yet it seems

s-urbaniak commented on examples/non-resource-url/non-resource-url/README.md:
```diff
@@ -0,0 +1,170 @@
+# non-resource-url example
```
same question as above: how do these example differ?

s-urbaniak commented:
@brancz looking good so far, just a couple of questions:
- the change commit contains vendoring code, do you mind to keep it separate?
- I _think_ there are duplicate readmes, or not?
- How does audiences relate to non-resource URL protection?

brancz commented:
> the change commit contains vendoring code, do you mind to keep it separate?

Done, sorry about that.

> I think there are duplicate readmes, or not?

Fixed this.

> How does audiences relate to non-resource URL protection?

It doesn't really, I just wanted to provide some example showing the functionality. I added a note at the top of the example readme to make sure that's clearer:

> Token audience reviews can be combined with any of the other features of kube-rbac-proxy, such as reviewing permission for specific resources, this example merely chooses to show the functionality using a non-resource-url permission.

metalmatze commented:
Are you still waiting for reviews or something else?

brancz commented:
Waiting for reviews :)

s-urbaniak commented on README.md:
````diff
@@ -28,18 +28,22 @@ All command line flags:
 ```txt
 $ kube-rbac-proxy -h
 Usage of _output/linux/amd64/kube-rbac-proxy:
+      --add_dir_header                              If true, adds the file directory to the header
````
I am wondering if this is used at all?

brancz replied:
It’s from klog

s-urbaniak commented on README.md:
```diff
@@ -48,15 +52,18 @@ Usage of _output/linux/amd64/kube-rbac-proxy:
       --tls-cipher-suites strings                   Comma-separated list of cipher suites for the server. Values are from tls package constants (https://golang.org/pkg/crypto/tls/#pkg-constants). If omitted, the default Go cipher suites will be used
       --tls-min-version string                      Minimum TLS version supported. Value must match version names from https://golang.org/pkg/crypto/tls/#pkg-constants. (default "VersionTLS12")
       --tls-private-key-file string                 File containing the default x509 private key matching --tls-cert-file.
+      --tls-reload-interval duration                The interval at which to watch for TLS certificate changes, by default set to 1 minute. (default 1m0s)
```
good catch :+1:

brancz replied:
This actually came in as I regenerated the help text

s-urbaniak commented on examples/non-resource-url-token-request/README.md:
```diff
@@ -0,0 +1,292 @@
+
+> Token audience reviews can be combined with any of the other features of kube-rbac-proxy, such as reviewing permission for specific resources, this example merely chooses to show the functionality using a non-resource-url permission.
+
+This examples is in essence similar to the [non-resource-url](../non-resource-url/) example, with the key difference, that this example requires the ServiceAccount token sent by a client must be scoped to the `kube-rbac-proxy.default.svc` audience. In this example the scoped ServiceAccount token is obtained via a projected volume and mounted into the client container from where it can be consumed ...
```
s/scroped/scoped

brancz replied:
Fixed.

brancz commented:
I split the PR into multiple commits, which should make it very easy to review now, as the actual code changes (in the second commit) are minimal and are mainly upgrading to using the v1 APIs. The rest are tests and docs (95% of this PR).

s-urbaniak commented on pkg/authn/delegating.go:
```diff
@@ -22,21 +22,33 @@ import (
+	if len(authn.X509.ClientCAFile) > 0 {
+		p, err = dynamiccertificates.NewStaticCAContentFromFile(authn.X509.ClientCAFile)
+		if err != nil {
+			return nil, err
```
just a nit: let's wrap this err, so context is visible in log lines.

s-urbaniak commented:
this lgtm :tada: I really just have one nit regarding error wrapping, else this looks great. A small code footprint, but great security improvement :+1:
